v0.1.5: fix manipulating multiple statements; support cross tab copy paste

HollowMan6 opened the pull request:
Resolve #8 

- Fix manipulating multiple statements by using `getSurroundParent` inside `hasSelectedParent` check when not in dragging.
- Fix some API names.
- Keep connection relationship for multiple statements after duplication / copy-then-paste.
- Support cross tab copy paste.

zakx commented:
Thank you very much @HollowMan6! 

I can confirm that duplicating multiple blocks using the context menu works correctly now.

However, when selecting multiple blocks and trying to duplicate them by using Cmd-C & Cmd-V on the keyboard, only the topmost block gets copied. Is that a related issue?

HollowMan6 commented:
Hi @zakx , I believe I had already fixed selecting multiple statements and trying to duplicate them by using Cmd-C & Cmd-V on the keyboard before. The only issue I found was that I mistakenly created a duplication in a reversed connection order when using Cmd-C & Cmd-V, I don't know if this caused you to think that only the topmost block gets copied, sorry about that, and I have just fixed this.

![GIF 2023-3-18 23-47-08](https://user-images.githubusercontent.com/43995067/226142587-ebff0d94-f24b-4db8-ab1a-3421be170f2c.gif)

zakx commented on src/multiselect_shortcut.js:
```diff
@@ -184,15 +198,19 @@ const registerCut = function() {
+          blockList.push(block.id);
+        }
+      };
+      const applyDelete = function(block) {
```
For some reason I couldn't figure out, applyDelete does get called with block === null for me, resulting in an error. 

Adding `if(!block) return;` here solved that issue for me and made cutting work correctly.

HollowMan6 replied:
Thank you!

zakx commented:
Hi @HollowMan6, thanks again for your swift response and help!

You are correct, copying and pasting via shortcuts do work correctly now. I had an issue with cutting though (see above).

It didn't work before because I also had `@blockly/plugin-cross-tab-copy-paste` in my installation and that handled the keyboard shortcuts -- apparently it has the same issue, I'll open another ticket there.

HollowMan6 commented:
> It didn't work before because I also had `@blockly/plugin-cross-tab-copy-paste` in my installation and that handled the keyboard shortcuts -- apparently it has the same issue, I'll open another ticket there.

I would say it's hard to integrate such plugins at the same time with the multiselect in this case, as we need to unregister the previous handling functions and replace that with ours to work, and they are basically doing the same thing. https://github.com/google/blockly-samples/blob/master/plugins/cross-tab-copy-paste/src/index.js#L41-L51

HollowMan6 commented:
I think the only feasible solution would be that we add the cross-tab-copy-paste feature into our plugin.

HollowMan6 commented:
Hi @zakx ! Now the cross-tab-copy-paste feature is available in our plugin. Any comments or suggestions are welcome!

/cc @ewpatton it would be great if you can help me review this PR and make a new release, so that the fix and new feature can be available asap, thank you!

ewpatton commented:
@HollowMan6 Is this stable now to review? You pushed one commit after tagging me to review it.

HollowMan6 commented:
Yes, it's stable now to review

ewpatton reviewed (changes requested):
There are some linting errors that come up when I try to build this with `npm run build`.

ewpatton commented:
@HollowMan6 `npm install` did not fix it but `npm ci` did.

ewpatton commented on src/multiselect_contextmenu.js:
```diff
@@ -9,7 +9,110 @@
+    },
+    preconditionFn: function(scope) {
+      const workspace = scope.block.workspace;
+      if (workspace.options.readOnly) {
+        return 'disabled';
+      }
```
Note that there is a discrepancy in behavior here with the paste option w.r.t. readOnly mode. Here, we return disabled, but in the paste option we return hidden. We probably also want this to be hidden (although with cross tab copy paste I think it could even be enabled since copying blocks doesn't modify the workspace).

HollowMan6 replied:
Thank you! I just changed this into that, for copy, we only return hidden when the workspace is read only and not in cross tab copy paste mode.

ewpatton commented on src/multiselect_contextmenu.js:
```diff
@@ -480,6 +595,80 @@ const registerDelete = function() {
+      if (blockNumGetFromStorage(useCopyPasteCrossTab) < 1) {
+        return 'disabled';
+      }
+      return !scope.workspace.options.readOnly?
+        'enabled': 'hidden';
```
It seems like we should always return hidden if the workspace is read-only, but as implemented if cross tab copy paste is enabled the state will either be disabled or hidden depending on whether there is content in the local storage. If it's disabled, that might imply to users that at some point it could be enabled, which will never be the case. Update this so that the read only check is first so that the option is hidden regardless of other factors.

HollowMan6 replied:
For paste, now it will always be hidden when the workspace is read-only, if it's not, then it depends on the content in the clipboard to decide whether it's enabled or disabled.

ewpatton reviewed (changes requested):
Two minor comments
